core/src/main/java/org/apache/lucene/analysis/miscellaneous/DuplicateByteSequenceSpotter.java

```diff
@@ -0,0 +1,325 @@
+                } else {
+                    // Deeper-level nodes are less visited but more numerous
+                    // so use a more space-friendly data structure
+                    node = new LightweightTreeNode(b, this, depth);
+                }
+                children[bIndex] = node;
+            }
+            return node;
+        }
+
+        @Override
+        public short add(int byteSequence) {
+            throw new UnsupportedOperationException("Root nodes do not support byte sequences encoded as integers");
+        }
+
+    }
+
+    // Node implementation for use by the depth 3 branches of the tree that
+    // sacrifices speed for space.
+    final class LightweightTreeNode extends TreeNode {
+
+        // An array dynamically resized ranging from 0 to 256 values but
+        // frequently only sized 1 as most sequences leading to end leaves are
+        // one-off paths
+        ByteSequenceLeafNode[] children = null;
+
+        LightweightTreeNode(byte key, TreeNode parentNode, int depth) {
+            super(key, parentNode, depth);
+            bytesAllocated += LIGHTWEIGHT_TREE_NODE_OBJECT_SIZE;
+
+        }
+
+        @Override
+        public short add(int byteSequence) {
+            if (children == null) {
+                // Create array adding new child
+                children = new ByteSequenceLeafNode[1];
+                bytesAllocated += RamUsageEstimator.NUM_BYTES_ARRAY_HEADER;
+
+                ByteSequenceLeafNode child = new ByteSequenceLeafNode(byteSequence);
+                bytesAllocated += RamUsageEstimator.NUM_BYTES_OBJECT_REF;
+                children[0] = child;
+                return child.hitCount;
+            }
+            // Find existing child and increment count
+            for (ByteSequenceLeafNode child : children) {
+                if (child.byteSequence == byteSequence) {
+                    if (child.hitCount < Short.MAX_VALUE) {
+                        child.hitCount++;
+                    }
+                    return child.hitCount;
+                }
+            }
+            // Grow array adding new child
+            ByteSequenceLeafNode[] newChildren = new ByteSequenceLeafNode[children.length + 1];
+            bytesAllocated += RamUsageEstimator.NUM_BYTES_OBJECT_REF;
+
+            System.arraycopy(children, 0, newChildren, 0, children.length);
+            children = newChildren;
+            ByteSequenceLeafNode child = new ByteSequenceLeafNode(byteSequence);
+            children[newChildren.length - 1] = child;
+            nodesResizedByDepth++;
+            return child.hitCount;
+        }
+
+        @Override
+        public TreeNode add(byte b, int depth) {
+            throw new UnsupportedOperationException("Leaf nodes do not take byte sequences");
+        }
+
+    }
+
+    // Leaf node implementation that represents a sequence of bytes as an int
+    // for a compact representation
+    class ByteSequenceLeafNode {
```

Comment:
There are masses of leaf nodes so I think the 255 max-precision on `hitCount` would be the right trade off. I expect most clients are not interested in accurate numbers for numbers of repetitions - perhaps even only that the number of repetitions is non-zero.